# Numpy Problems

## Find Prime numbers between 1 and 1000, (use only numpy array and slicing etc)

### Notes
numpy has __vectorize__ function just like __map__ python function
```python
import numpy as np

vfunc = np.vectorise(lambda x, y: x % y == 0)
vfunc(np.array([2,3,4,5,6,7]),2)
```
Above example vfun returns array, *vfunc is much like npsin/npcos. etc

In [1]:
import numpy as np

In [2]:
def is_prime(n):
    if n % 2 == 0 and n > 2: 
        return False
    return all(n % i for i in range( 1, 1000)
     a = np.arange(1, 1000)
abc = np.vectorize(is_prime)
abc_d = abc(a)
primes = np.extract(abc_d, a)

SyntaxError: invalid syntax (<ipython-input-2-c20aa1449062>, line 5)

In [3]:
arr1 = np.arange(1,1000)

In [4]:
arr1

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104,
       105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
       131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
       144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156,
       157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169,
       170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 18

In [5]:
np.vectorize(arr1)

In [6]:
vfunc = np.array([arr1])

In [7]:
vfunc

array([[  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
         14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
         27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
         40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
         53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
         66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
         79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
         92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104,
        105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
        118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
        131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
        144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156,
        157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169,
        170, 171, 172, 173, 174, 175, 176, 177, 178

In [8]:
np.array

<function numpy.core.multiarray.array>

In [9]:
arr1[2:3:5]

array([3])

In [10]:
arr1[2:3]

array([3])

In [11]:
arr1[:]

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104,
       105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
       131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
       144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156,
       157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169,
       170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 18

In [12]:
arr1[:3]

array([1, 2, 3])

In [13]:
arr1[8:]

array([  9,  10,  11,  12,  13,  14,  15,  16,  17,  18,  19,  20,  21,
        22,  23,  24,  25,  26,  27,  28,  29,  30,  31,  32,  33,  34,
        35,  36,  37,  38,  39,  40,  41,  42,  43,  44,  45,  46,  47,
        48,  49,  50,  51,  52,  53,  54,  55,  56,  57,  58,  59,  60,
        61,  62,  63,  64,  65,  66,  67,  68,  69,  70,  71,  72,  73,
        74,  75,  76,  77,  78,  79,  80,  81,  82,  83,  84,  85,  86,
        87,  88,  89,  90,  91,  92,  93,  94,  95,  96,  97,  98,  99,
       100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112,
       113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125,
       126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138,
       139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151,
       152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164,
       165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177,
       178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 18

In [14]:
arr2 = (arr1 > 20) & (arr1 < 500)

In [15]:
arr2

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,

In [16]:
arr1[arr1%5==0]

array([  5,  10,  15,  20,  25,  30,  35,  40,  45,  50,  55,  60,  65,
        70,  75,  80,  85,  90,  95, 100, 105, 110, 115, 120, 125, 130,
       135, 140, 145, 150, 155, 160, 165, 170, 175, 180, 185, 190, 195,
       200, 205, 210, 215, 220, 225, 230, 235, 240, 245, 250, 255, 260,
       265, 270, 275, 280, 285, 290, 295, 300, 305, 310, 315, 320, 325,
       330, 335, 340, 345, 350, 355, 360, 365, 370, 375, 380, 385, 390,
       395, 400, 405, 410, 415, 420, 425, 430, 435, 440, 445, 450, 455,
       460, 465, 470, 475, 480, 485, 490, 495, 500, 505, 510, 515, 520,
       525, 530, 535, 540, 545, 550, 555, 560, 565, 570, 575, 580, 585,
       590, 595, 600, 605, 610, 615, 620, 625, 630, 635, 640, 645, 650,
       655, 660, 665, 670, 675, 680, 685, 690, 695, 700, 705, 710, 715,
       720, 725, 730, 735, 740, 745, 750, 755, 760, 765, 770, 775, 780,
       785, 790, 795, 800, 805, 810, 815, 820, 825, 830, 835, 840, 845,
       850, 855, 860, 865, 870, 875, 880, 885, 890, 895, 900, 90

In [17]:
arr1[arr1%2==0]

array([  2,   4,   6,   8,  10,  12,  14,  16,  18,  20,  22,  24,  26,
        28,  30,  32,  34,  36,  38,  40,  42,  44,  46,  48,  50,  52,
        54,  56,  58,  60,  62,  64,  66,  68,  70,  72,  74,  76,  78,
        80,  82,  84,  86,  88,  90,  92,  94,  96,  98, 100, 102, 104,
       106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130,
       132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156,
       158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182,
       184, 186, 188, 190, 192, 194, 196, 198, 200, 202, 204, 206, 208,
       210, 212, 214, 216, 218, 220, 222, 224, 226, 228, 230, 232, 234,
       236, 238, 240, 242, 244, 246, 248, 250, 252, 254, 256, 258, 260,
       262, 264, 266, 268, 270, 272, 274, 276, 278, 280, 282, 284, 286,
       288, 290, 292, 294, 296, 298, 300, 302, 304, 306, 308, 310, 312,
       314, 316, 318, 320, 322, 324, 326, 328, 330, 332, 334, 336, 338,
       340, 342, 344, 346, 348, 350, 352, 354, 356, 358, 360, 36

In [18]:
primes = np.array(arr1)

In [19]:
idx = arr1[arr1%2 == 0]

In [20]:
idx

array([  2,   4,   6,   8,  10,  12,  14,  16,  18,  20,  22,  24,  26,
        28,  30,  32,  34,  36,  38,  40,  42,  44,  46,  48,  50,  52,
        54,  56,  58,  60,  62,  64,  66,  68,  70,  72,  74,  76,  78,
        80,  82,  84,  86,  88,  90,  92,  94,  96,  98, 100, 102, 104,
       106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130,
       132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156,
       158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182,
       184, 186, 188, 190, 192, 194, 196, 198, 200, 202, 204, 206, 208,
       210, 212, 214, 216, 218, 220, 222, 224, 226, 228, 230, 232, 234,
       236, 238, 240, 242, 244, 246, 248, 250, 252, 254, 256, 258, 260,
       262, 264, 266, 268, 270, 272, 274, 276, 278, 280, 282, 284, 286,
       288, 290, 292, 294, 296, 298, 300, 302, 304, 306, 308, 310, 312,
       314, 316, 318, 320, 322, 324, 326, 328, 330, 332, 334, 336, 338,
       340, 342, 344, 346, 348, 350, 352, 354, 356, 358, 360, 36

In [21]:
np.add(np.array([1,2,3]),2)

array([3, 4, 5])

In [22]:
np.array([1,2,3]) + 2

array([3, 4, 5])

In [23]:
np.add.reduce(np.array([1,2,3]))

6

### any and all

In [24]:
np.any(np.array([True,False,True]))

True

In [25]:
np.all(np.array([True,False,True]))

False

In [26]:
np.all(np.array([True,True,True]))

True

when one element is taken the element behind it should be sliced

In [27]:
arr1 = np.arange(1,1000)

In [28]:
arr1

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104,
       105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
       131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
       144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156,
       157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169,
       170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 18

In [29]:
np.any(np.array([arr1]))

True

In [30]:
np.all(np.array([arr1]))

True

In [31]:
np.all(~np.array([False,False,False]))

True

In [32]:
import numpy as np
import math

def is_prime(n):
    if n % 2 == 0 and n > 2: 
        return False
    return all(n % i for i in range(3, int(math.sqrt(n)) + 1, 2))

a = np.arange(1, 10**3)
foo = np.vectorize(lambda x: (not (x + 1) % 5 or not (x - 1) % 5) and is_prime(x))
primes = a[foo(a)]
print(primes)

[  1  11  19  29  31  41  59  61  71  79  89 101 109 131 139 149 151 179
 181 191 199 211 229 239 241 251 269 271 281 311 331 349 359 379 389 401
 409 419 421 431 439 449 461 479 491 499 509 521 541 569 571 599 601 619
 631 641 659 661 691 701 709 719 739 751 761 769 809 811 821 829 839 859
 881 911 919 929 941 971 991]


Problem 25
The Fibonacci sequence is defined by the recurrence relation:

Fn = Fn−1 + Fn−2, where F1 = 1 and F2 = 1.
Hence the first 12 terms will be:

F1 = 1,
F2 = 1,
F3 = 2,
F4 = 3,
F5 = 5,
F6 = 8,
F7 = 13,
F8 = 21,
F9 = 34,
F10 = 55,
F11 = 89,
F12 = 144,
The 12th term, F12, is the first term to contain three digits.

What is the index of the first term in the Fibonacci sequence to contain 1000 digits?

In [33]:
def fibo(n):
    a = 1
    b = 0
    while True :
        return a
    a,b = a, a+b
    

In [34]:
len(list(str(12786543)))

8

In [35]:
len(str(128786543))

9

In [36]:
int(np.log10(128786543)+1)

9

In [37]:
np.log10(np.float64(123456789009876543))

17.091514977204014

In [38]:
np.int64

numpy.int64

In [39]:
np.log10(np.float64(123456789009875468720238903746748292307236543))

44.091514977204007

In [40]:
a = 1
b = 1
result = 2
while len(str(b)) < 1000:
    a, b = b, a+b
    result += 1
print(result)

4782


In [41]:
int(567).log10

AttributeError: 'int' object has no attribute 'log10'

In [42]:
a = np.int64(567)
a.log10

AttributeError: 'numpy.int64' object has no attribute 'log10'

In [43]:
b= np.float64(567)
a.log10

AttributeError: 'numpy.int64' object has no attribute 'log10'

In [44]:
help(np.log10)

Help on ufunc object:

log10 = class ufunc(builtins.object)
 |  Functions that operate element by element on whole arrays.
 |  
 |  To see the documentation for a specific ufunc, use np.info().  For
 |  example, np.info(np.sin).  Because ufuncs are written in C
 |  (for speed) and linked into Python with NumPy's ufunc facility,
 |  Python's help() function finds this page whenever help() is called
 |  on a ufunc.
 |  
 |  A detailed explanation of ufuncs can be found in the "ufuncs.rst"
 |  file in the NumPy reference guide.
 |  
 |  Unary ufuncs:
 |  =============
 |  
 |  op(X, out=None)
 |  Apply op to X elementwise
 |  
 |  Parameters
 |  ----------
 |  X : array_like
 |      Input array.
 |  out : array_like
 |      An array to store the output. Must be the same shape as `X`.
 |  
 |  Returns
 |  -------
 |  r : array_like
 |      `r` will have the same shape as `X`; if out is provided, `r`
 |      will be equal to out.
 |  
 |  Binary ufuncs:
 |  ==============
 |  
 |  op(X, Y, 

In [45]:
np.log10(567)

2.7535830588929064

In [46]:
np.log10([567])

array([ 2.75358306])

In [47]:
arr = np.array([1,2])

In [49]:
np.info(np.append)

 append(arr, values, axis=None)

Append values to the end of an array.

Parameters
----------
arr : array_like
    Values are appended to a copy of this array.
values : array_like
    These values are appended to a copy of `arr`.  It must be of the
    correct shape (the same shape as `arr`, excluding `axis`).  If
    `axis` is not specified, `values` can be any shape and will be
    flattened before use.
axis : int, optional
    The axis along which `values` are appended.  If `axis` is not
    given, both `arr` and `values` are flattened before use.

Returns
-------
append : ndarray
    A copy of `arr` with `values` appended to `axis`.  Note that
    `append` does not occur in-place: a new array is allocated and
    filled.  If `axis` is None, `out` is a flattened array.

See Also
--------
insert : Insert elements into an array.
delete : Delete elements from an array.

Examples
--------
>>> np.append([1, 2, 3], [[4, 5, 6], [7, 8, 9]])
array([1, 2, 3, 4, 5, 6, 7, 8, 9])

When `axis` is

In [50]:
arr[-2:]

array([1, 2])

In [56]:
arr = np.append(arr, np.sum(arr[-2:]))

In [57]:
arr = np.append(arr, np.sum(arr[-2:]))

In [58]:
arr

array([1, 2, 3, 5, 8])

In [59]:
arr = np.array([1,2])

In [60]:
arr = np.array([arr[-1]], np.sum(arr))

In [61]:
arr

array([2])

In [65]:
arr = np.array([1,2])
while np.log10(np.float64(arr[-1])) < 10:
    arr = np.array([arr[-1],np.sum(arr)])
print(arr)

[ 7778742049 12586269025]
